granadilla/forms.py

```diff
@@ -48,6 +49,116 @@ class Meta:
+                return True
+        return False
+
+    def save(self, commit=True):
+        device = super(LdapDeviceForm, self).save(False)
+
+        device.device_username = self.__username
+        user = models.LdapUser.objects.get(username=device.device_username)
+        device.device_owner = user.dn
+        device.device_fullname = device.device_username \
+                         + '_' + self.data['device_name']
+
+        device.set_password(self.data['password'])
+        if self.__base_dn:
+            device.base_dn = self.__base_dn
+        if commit:
+            device.save()
+        return device
+
+    class Meta:
+        model = LdapDevice
+        widgets = {
+            'password': forms.PasswordInput(),
+        }
+        exclude = ("dn","device_username", "device_owner", "device_fullname")
+
+class PasswordForm(forms.ModelForm):
+    oldpass = forms.CharField(widget=forms.PasswordInput, label = "Old Password")
+    password1 = forms.CharField(widget=forms.PasswordInput, label = "New Password")
+    password2 = forms.CharField(widget=forms.PasswordInput, label = "Re-Password")
+    
+    def __init__(self, *args, **kwargs):
+        if 'base_dn' in kwargs:
+            self.__base_dn = kwargs.pop('base_dn')
+        else:
+            self.__base_dn = None
+        self.__fullname = kwargs.pop('device_fullname')
+        super(PasswordForm, self).__init__(*args, **kwargs)
+
+
+    def is_valid(self):
+        if super(PasswordForm, self).is_valid():    
+            try:
+                device = models.LdapDevice.objects.get(device_fullname=self.__fullname)
+                # Check if the old device is valid
+                if device.check_password(self.data['oldpass']):
+                    # Check if the two new passwords are the same
+                    password1 = self.data['password1'] 
+                    password2 = self.data['password2'] 
+                    if password1 and password1 != password2: 
+                        self.errors['passworderror'] = "Passwords don't match" 
+                        return False 
+                    return True
+                else:
+                    self.errors['wrong password'] = "The old password is wrong"
+                    return False
+            except models.LdapDevice.DoesNotExist:
+                self.errors['wrong device'] = "This device does not exist"
+                return False
+        return False
+
+    def save(self, commit=True):
+        device = models.LdapDevice.objects.get(device_fullname=self.__fullname)
+
+        device.set_password(self.data['password1'])
+        if commit:
+            device.save()
+        return device
+
+    class Meta:
+        model = LdapDevice
+        widgets = {
+            'oldpass': forms.PasswordInput(),
+            'password1': forms.PasswordInput(),
+            'password2': forms.PasswordInput(),
+        }
+        exclude = ("dn","device_username", "device_name", "device_owner",\
```

Comment:
Since you're excluding so many fields, wouldn't a basic ModelForm be more suitable?